### PR TITLE
[RPC] Provider-Register transactions

### PR DESCRIPTION
**Pull request by random-zebra**
Extracted from #2267. Based on top of
- [x] #2273

Implement `protx_register` and `protx_list` RPC commands. 
Add DMN support to `listmasternodes`.
Deprecate `startmasternode` after SPORK_21 activation.

### CODE REVIEW

**Comment by random-zebra** [comment]
Rebased on master. This too is ready to go.

**Review by furszy — DISMISSED** [review]
Left some comments mostly over the new commands help texts. Other than that code review ACK b0899bce84fb21b1493389fb891b34f5dc9f7f08 .

There are some architectural improvements that we can do moving forward but nothing that would stop this for getting merge. Better to have a high percentage of test coverage before start with them.

**Comment by random-zebra** [comment]
Help texts fixed as per feedback. RPC-Cli examples will be added later (ref https://github.com/PIVX-Project/PIVX/pull/2267/commits/2f71d8d2837ab89aa4a223c8c682f1255fd441ee).

**Review by furszy — COMMENTED** [review]
Small finding, cherry-pick this one here if you want a00adc5293b96214800ba58d9ebae75eb66ab1df

We missed to lock the Masternode internal mutex when the collateral spent flag is set. Which could cause a concurrency problem between the message handler thread and the tier two thread as they are both accessing to the same data.

Other than that, no functional changes after the help text fixes and the rebase.
All good and ready to go for me 👍

**Comment by random-zebra** [comment]
Commit cherry-picked.

**Review by furszy — APPROVED** [review]
code ACK e754732d87a24450fb7f0a1c86a1e58d1a83ac1d after cherry-pick.

File: src/rpc/rpcevo.cpp [hunk 1]
```diff
@@ -0,0 +1,764 @@
...
+}
+
+static std::string TxInErrorToString(int i, const CTxIn& txin, const std::string& strError)
+{
+    return strprintf("Input %d (%s): %s", i, txin.prevout.ToStringShort(), strError);
+}
+
+static OperationResult SignTransaction(CMutableTransaction& tx)
+{
+    EnsureWalletIsUnlocked();
```
Comment (Fuzzbawls) [inline src/rpc/rpcevo.cpp:277]: nit: redundant check

File: src/rpc/rpcevo.cpp [hunk 2]
```diff
@@ -0,0 +1,764 @@
...
+        }
+        UpdateTransaction(tx, i, sigdata);
+    }
+    return OperationResult(true);
+}
+
+static std::string SignAndSendSpecialTx(CMutableTransaction& tx, const ProRegPL& pl)
+{
+    EnsureWallet();
+    EnsureWalletIsUnlocked();
```
Comment (Fuzzbawls) [inline src/rpc/rpcevo.cpp:301]: nit: redundant checks

**Review by Fuzzbawls — APPROVED** [review]
Code ACK e754732d87a24450fb7f0a1c86a1e58d1a83ac1d

With some nits for redundant checks to ensure the wallet exists or is unlocked. these checks are done directly inside the main RPC command functions and shouldn't need to be re-done in dependent helper functions.

Note: the `SignSpecialTxPayloadByHash()` and `ParsePrivKey()` functions are also included here unused, but connected in a separate PR, which i think is fine.

Also, for future discussion prior to 6.0 release: as with the sapling RPC command names when they were first introduced; I'd really like to avoid snake case in our final naming convention, but that is not something to change here and can be discussed in a separate issue without slowing down the progress of this series of PRs.

**Comment by random-zebra** [comment]
> With some nits for redundant checks to ensure the wallet exists or is unlocked.

Good catch.
I'm going to fix this inside #2308 (which is the next one in line), so the reviews here don't get dismissed, and we can move forward.

> as with the sapling RPC command names when they were first introduced; I'd really like to avoid snake case in our final naming convention

Yes, I thought so.
But `protxregisterprepare`, for example, seemed rather ugly. In any case, yeah, we can discuss this, and update it with a scripted-commit, before the final release.

Merging...

**Comment by Fuzzbawls** [comment]
yes, the particular naming convention is something completely outside the scope of these PRs, which is why i only mentioned it as an aside comment and not a blocking comment...has nothing to do with the code itself, as the command names can easily be changed prior to release without issue.

As more RPC commands are added in this chain of PRs, I'm thinking it may be best for me to open up an RFC issue specifically for the purpose of brainstorming/discussing more appropriate command names that conform to our existing naming convention, that won't have any tangible effect/block on existing PRs.

**Comment by random-zebra** [comment]
> I'm thinking it may be best for me to open up an RFC issue specifically for the purpose of brainstorming/discussing more appropriate command names that conform to our existing naming convention

Sounds good.
Btw, RPC redundant checks removed in https://github.com/PIVX-Project/PIVX/pull/2308/commits/c01639bb9255644e2f28e7620b1ec2f1e8aa0d2b.

**Pull request merged**